src/redecomp/transfer/SparseMatrixTransfer.cpp

```diff
@@ -0,0 +1,321 @@
+      for (auto& send_col_offsets : send_col_offsets_by_rank)
+      {
+        send_col_offsets.reserve(approx_rank_size);
+      }
+      for (int i{0}; i < send_parent_local_col_offsets[r].size(); ++i)
+      {
+        send_col_offsets_by_rank[send_parent_col_ranks[r][i]].push_back(send_parent_local_col_offsets[r][i]);
+      }
+      auto recv_col_offsets_by_rank = MPIArray<int>(&getMPIUtility());
+      recv_col_offsets_by_rank.SendRecvArrayEach(send_col_offsets_by_rank);
+      // convert (now on-rank) offsets to element ids
+      for (int trial_r{0}; trial_r < n_ranks; ++trial_r)
+      {
+        for (auto& recv_col_offset : recv_col_offsets_by_rank[trial_r])
+        {
+          recv_col_offset = redecomp_trial_mesh_.getParentToRedecompElems().first[trial_r][recv_col_offset];
+        }
+      }
+      // send back to original rank
+      send_col_offsets_by_rank.SendRecvArrayEach(recv_col_offsets_by_rank);
+      // fill send_parent_local_col_offsets with local element ids
+      axom::Array<int> per_rank_ct(n_ranks, n_ranks);
+      for (int i{0}; i < send_parent_local_col_offsets[r].size(); ++i)
+      {
+        auto col_rank = send_parent_col_ranks[r][i];
+        send_parent_local_col_offsets[r][i] = send_col_offsets_by_rank[col_rank][per_rank_ct[col_rank]++];
+      }
+    }
+    // do MPI communication
+    recv_matrix_data.SendRecvArrayEach(send_matrix_data);
+    // NOTE: these hold offsets now; need to convert these to element numbers
+    recv_parent_local_row.SendRecvArrayEach(send_parent_local_row_offsets);
+    // NOTE: the send values were converted from offsets to element numbers in the block above
+    recv_parent_local_col.SendRecvArrayEach(send_parent_local_col_offsets);
+    recv_parent_col_ranks.SendRecvArrayEach(send_parent_col_ranks);
+    // convert local row offsets to element numbers
+    for (int r{0}; r < n_ranks; ++r)
+    {
+      for (auto& row_elem : recv_parent_local_row[r])
+      {
+        row_elem = redecomp_test_mesh_.getParentToRedecompElems().first[r][row_elem];
+      }
+    }
+  }
+
+  // Now we need to build the CSR data for the hypre diag and offd matrices.  diag holds data with both rows and cols
+  // on-rank and offd holds data with rows on-rank and cols off-rank.
+  auto col_starts = BuildParentElementRankOffsets(redecomp_trial_mesh_);
+  // size i_diag and i_offd, number of columms per row of matrix, and build a global dof to local offd column map
+  auto diag_nnz = 0;
+  auto offd_nnz = 0;
+  // NOTE: these will be owned by the HypreParMatrix. using mfem::Memory, the data will not be deleted upon destruction.
+  mfem::Memory<HYPRE_Int> i_diag(parent_test_space_.GetVSize() + 1);
+  mfem::Memory<HYPRE_Int> i_offd(parent_test_space_.GetVSize() + 1);
+  for (int i{0}; i < parent_test_space_.GetVSize() + 1; ++i)
+  {
+    i_diag[i] = 0;
+    i_offd[i] = 0;
+  }
+  // maps from global j to order of appearance in the loops below
```

Comment:
Consider clarity of comment. "in the loops below" isn't super clear. Global j may be hard to parse for someone not immediately refreshed on CSR data. This is complex code to be sure, but I'm not sure that this comment enhances understanding of the next handful of lines.